### ISSUE TITLE
Email as a common name.

### ISSUE THREAD

**Issue by evgenii-mazhora** [jkroepke/openvpn-auth-oauth2#100]
### Problem Statement

Hello.
I'm trying to set up a plugin with Google OIDC, and I have a small question. Can I use the email from the data obtained from OIDC to fill in the common name field?
Let me try to explain in more detail. I am using a configuration without client certificates and passwords (a sample client config is below). All authentication is performed only through openvpn-auth-oauth2. This is quite convenient, one configuration file for all clients.

```
remote vpn.example.com 8001
remote-cert-tls server
client
dev tun
proto tcp
key-direction 1
cipher AES-256-CBC
nobind
persist-key
verb 3
keepalive 10 60
 
<ca>
.....
</ca>
<tls-auth>
.....
</tls-auth>
```

But in this case, there is a small problem, since we do not specify a username anywhere, which can later be used as CN, so if you look at the statistics, it is not clear which user is which client (Common Name = UNDEF)

```
root@localhost:~# cat  /var/run/openvpn-server/status-tcp8001.log
OpenVPN CLIENT LIST
Updated,2023-12-27 08:10:21
Common Name,Real Address,Bytes Received,Bytes Sent,Connected Since
UNDEF,219.93.0.130:48881,1784,3808,2023-12-27 08:10:02
UNDEF,219.93.0.130:58579,1558,3660,2023-12-27 08:10:16
ROUTING TABLE
Virtual Address,Common Name,Real Address,Last Ref
10.8.0.36,UNDEF,219.93.0.130:58579,2023-12-27 08:10:19
10.8.0.37,UNDEF,219.93.0.130:48881,2023-12-27 08:10:20
GLOBAL STATS
Max bcast/mcast queue length,2
END
root@localhost:~#
```

Can you advise how to configure the plugin so that it passes the CN value to OpenVPN through the management interface, or is such functionality not available at the moment?

**Comment by jkroepke** [comment]
Hi @evgenii-mazhora 

> or is such functionality not available at the moment?

The OpenVPN server does not have the capability to override any client field.

I already request the feature month ago: https://github.com/OpenVPN/openvpn/issues/299

You can also leave some feedback there, including your use-case. If more people proving use-cases, there is a chance that the feature is included in the next minor version.

**Comment by aprudnev** [comment]
It looks as OVPN can use user name instead of common name. But how can we map it? Here is what I see in the logs when I run our test:

Feb 26 21:20:46 stg5testovpnw01 openvpn-auth-oauth2[3663]: time=2024-02-26T21:20:46.396-08:00 level=INFO msg="accept OpenVPN client cid 56, kid 1" cid=56 kid=1 common_name=EIS-CLIENT idtoken.subject=aroudnev idtoken.email="" idtoken.preferred_username="" user.subject=aroudnev user.preferred_username=""
Feb 26 21:20:46 stg5testovpnw01 ovpn-server[3670]: MANAGEMENT: CMD 'client-auth 56 1'
Feb 26 21:20:46 stg5testovpnw01 ovpn-server[3670]: 73.15.8.25:49852 PUSH: Received control message: 'PUSH_REQUEST'
Feb 26 21:20:46 stg5testovpnw01 openvpn-auth-oauth2[3663]: time=2024-02-26T21:20:46.743-08:00 level=INFO msg="client established" cid=56 common_name="" reason=ESTABLISHED
Feb 26 21:20:46 stg5testovpnw01 ovpn-server[3670]: 73.15.8.25:49852 MULTI_sva: pool returned IPv4=10.25.25.106, IPv6=(Not enabled)
Feb 26 21:20:46 stg5testovpnw01 ovpn-server[3670]: 73.15.8.25:49852 MULTI: Learn: 10.25.25.106 -> 73.15.8.25:49852
Feb 26 21:20:46 stg5testovpnw01 ovpn-server[3670]: 73.15.8.25:49852 MULTI: primary virtual IP for 73.15.8.25:49852: 10.25.25.106
Feb 26 21:20:46 stg5testovpnw01 ovpn-server[3670]: 73.15.8.25:49852 SENT CONTROL []: 'PUSH_REPLY,dhcp-option DNS 10.21.5.31,dhcp-option DNS 10.23.5.31,dhcp-option DNS 192.168.8.24,redirect-gateway def1 bypass-dhcp,route 10.0.0.0 255.0.0.0,route 192.168.0.0 255.255.0.0,route 172.16.0.0 255.240.0.0,reneg-sec 0,route-gateway 10.25.25.105,topology subnet,ping 10,ping-restart 120,auth-token-user YXJvdWRuZXY=,ifconfig 10.25.25.106 255.255.255.248,peer-id 0,auth-tokenSESS_ID,cipher AES-128-GCM,protocol-flags cc-exit tls-ekm dyn-tls-crypt,tun-mtu 1500' (status=1)
Feb 26 21:20:47 stg5testovpnw01 ovpn-server[3670]: 73.15.8.25:49852 Data Channel: cipher 'AES-128-GCM', peer-id: 0


And, earlier, when I did not add an option for the username, OVPN reported client by CN . Co it looks as all we need to do is to map common name:

Feb 26 18:58:48 stg5testovpnw01 ovpn-server[3590]: EIS-CLIENT/166.170.40.159:47528 SENT CONTROL [EIS-CLIENT]: 'PUSH_REPLY,dhcp-option DNS 10.21.5.31,dhcp-option DNS 10.23.5.31,dhcp-option DNS 192.168.8.24,redirect-gateway def1 bypass-dhcp,route 10.0.0.0 255.0.0.0,route 192.168.0.0 255.255.0.0,route 172.16.0.0 255.240.0.0,reneg-sec 0,route-gateway 10.25.25.105,topology subnet,ping 10,ping-restart 120,auth-token-user YWxleGVpcg==,ifconfig 10.25.25.107 255.255.255.248,peer-id 1,auth-tokenSESS_ID,cipher AES-128-GCM,key-derivation tls-ekm' (status=1)

(EIS-CLIENT is CN from the certificate)

**Comment by aprudnev** [comment]
One more problem we discover once we run POC is that I can not see, how persistent session is configured and, more important, how can we reset session tokens on disconnect. IT is known problem (that logouts are way more difficult vs logins, in any SSO) but it persist somehow. It can be great if I can configure kind of idle lifetime or a way to say _if user disconnects then system should invalidate SSO session_. Maybe I can configure it on IdP side, I just used existing OIDC configuration and I have a huge number of options with both PingFederate and PingOne (and we can use any), so I hope to find a solution. One method can be to customize login (access granted) web page, adding SSO logout into it, other can be better understanding of persistent configuration options. We just started (I made test system running few hours ago) and it all looks very promising.

**Comment by jkroepke** [comment]
> how can we reset session tokens on disconnect

By default, session tokens will not used across disconnects, unless `oauth2.refresh.use-session-id=true` is configured.

If supported by SSO, openvpn-auth-oauth2 will sent the SSO refresh_token to the revoke endpoint of the SSO provider to make the token invalid.

You can configure idle time with 

* `oauth2.refresh.expires` The timer will be reset on each REAUTH.

**Comment by aprudnev** [comment]
Good idea, I will test it.

**Comment by aprudnev** [comment]
Actually, does openvpn send IP and other information back to the plugin? If
it sends it, plugin can log IP with other information itself (we need -
user name, client-s IP, assigned IP if any was assigned, start and end of
the session with this IP).

The reason is simple. Let's say I (as NOC) receive a complaint from - say
our provider - dear XXX, we receive a complaint from YYY media company that
a copyrighted file was downloaded (exact time, IP , port, protocol, time of
session). Please investigate. (+ usual legal bullshit).

Our policy requires us to log all VPN sessions in a central place, and it
includes user,IP, external IP, connect/disconnect times and number of bytes.

So we first check on firewall logs who was NATed to this IP and port in
time of violation. We keep all firewall logs including NAT events so it is
easy. Then we check who was under this IP in time of violation, using VPN
logs (actually we integrate all logs but the idea is the same). And, then,
we know the exact person who did it (Yes, it is simplification, as a common
VPN user can not download Disney movies - many protections and FW rules
will prevent it - but the idea is the same for other events, up to 'post
files by mistake to the wrong repo'.

And this is why it is absolutely critical to have such logs in a single
record. Of course we can see logs all together and can guess that 'this
auth log matches this vpn log, just by time' but it can not eliminate race
conditions etc. And this is why I added this requirement to company-s
policy many years ago, and never regret doing it.

**Comment by jkroepke** [comment]
Yup, I will add the client IP to the logs.

https://github.com/jkroepke/openvpn-auth-oauth2/pull/199

Any future information then can be lookup by client IP from OpenVPN itself.

**Comment by aprudnev** [comment]
Yes, please.  It's maybe is an easiest thing to do and can be enough to
cross link logs.

If you can see IP assigned to the client, add it too (but I have doubts
about it as IP assignment is after authentication). As I said, ideally
(which I never saw) we need log of

Client-s IP, Client-s name, Operation (connected; assigned IP IP,
reconnected, disconnected). Any IP (even client-s IP) will work (with a
very rare exceptions) as it allows to link together OVPN logs and plugin
logs. OVPN log client-s IP in all (almost) messages so it is good cross
reference for ovpn/oauth logs.

**Comment by aprudnev** [comment]
Ok.

All your advices regarding refresh token works well.
I got login page template from your code (actually it can be much better if
you have it as separate file as an example) and adjusted to our
requirements (so we show a way to end session now).

I can post information about working configuration to PingID community, if
you want.

**Comment by jkroepke** [comment]
> I can post information about working configuration to PingID community, if
you want.

Yes, please.

It would be also awesome, if you could provide a Guide, what needs to be configured at PingID. Like we have it for other providers:

https://github.com/jkroepke/openvpn-auth-oauth2/wiki/Providers

**Comment by aprudnev** [comment]
I will see if I can do  it. PingID is definitely well compatible with your
system, BUT Ping has many possible configurations - you can use PingOne or
PingFederate as IdP, and PingOne can be PingOne or PingOneForEnterprise.

I am going to create internal document _HOW TO SET IT UP_ and I will
provide some extracts from it and obfuscated configurations for your
documentation, too.

PS. It is maybe #3 POC (PROOF OF CONCEPT) of related things. Most worked as
_yes, it can be configured, and it works, but looks not stable or looks as
it will break on next update_. This combination (Ubuntu Linux, OpenVPN ,
your plugin, and PingFederate as IdP), works extremely stable and we plan
to add it into the production as addition to existing VPN systems, which
uses RADIUS and so have serious limitations, for example when we run RADIUS
and MFA, you can not use SMS or EMAIL or FIDO devices as a second factor.
WebAuth allows it all.

**Comment by jkroepke** [comment]
Thanks a lot!

Yeah, the legacy stuff (RADIUS, OAM, LDAP) doesnt work with modern authentication system and that was the root cause for openvpn-auth-oauth2.

**Comment by aprudnev** [comment]
PingID and openvpn-auth-oauth2 - document is here

https://docs.google.com/document/d/10na-NYZ-EamlufxO6sBuWbu9WQkSmj5Y/edit?usp=sharing&ouid=104011126278844356239&rtpof=true&sd=true

It is a public version of our document, I clean it from all possible
company related things (and it is released for the open source, as we have
few of our own projects on github, like open source VTI router). You can
copy it, convert to wiki, simplify, etc, as you wish. Just let me know if
it is useful and how you manage to use it. If you prefer to keep it as it
is now, it's ok too (It is accessible by the link).

Issues I had with documentation during POC / installation
- it was not clean that redirect uri is /oauth2/callback (and I was lucky
that I used test setup  and it already had it with wildcard domain)
- it was not obvious, how to find list of options (yaml file is not full,
and list of command options is not in clear view)
- template for the message page - option do exists, but template example do
not (I extracted it from binary code using 'strings' command :) )
- when using web browser (when auth is provided by external browser), it is
important to understand that if I run few chrome profiles, then it opens
auth window in the currently used browser instance. It may cause big
confusion - I had one profile when I started VPN, then I change to another
one, and when VPN client try to re-negotiate it opens new session in new
chrome browser instance, not knowing about old old, and can not reconnect
at all. It's not a problem but only if you expect it.

**Comment by jkroepke** [comment]
> - it was not obvious, how to find list of options (yaml file is not full,
and list of command options is not in clear view)

Could you please do an example, what is missing?

> - template for the message page - option do exists, but template example do
not (I extracted it from binary code using 'strings' command :) )

You mean this one? https://github.com/jkroepke/openvpn-auth-oauth2/blob/main/internal/ui/index.gohtml

**Comment by aprudnev** [comment]
2. Yes, this one. Maybe, it make sense to have it as a file or have it in
documentation (and yes, I could take it from github, not a problem). I had
big question _how can user end session_ (as just disconnect do not end it,
and I am not sure how to configure LOGOUT options in PF which is way too
complicated for an easy experiments) I just added logout and dock links
into this web page.

1. Not much, but I do not see for example

# /usr/bin/openvpn-auth-oauth2 --help | grep common-name-case-sensitive
  --oauth2.validate.common-name-case-sensitive

BUT

# grep common-name-case-sensitive /etc/openvpn-auth-oauth2/*yaml
#

Same for socket-mode.

Few more maybe.  So I got impression that yaml file is not exactly up to
date, PLUS it do not contain default values. Usually (see sshd config for
example) config file has default values, AND separate example may have
something else.

**Comment by aprudnev** [comment]
SO far, I created our multi service system (3 services on one Linux -
stable fast full) with embedded web proxy and automated scripts.

Works very well. SO the last thing I expect to be fixed is _client-s IP in
AUTH logs and/or user id in openvpn logs_ . It's not a show stopper but is
important.

I will see if maybe I can port scripts etc as our open source project.

**Comment by jkroepke** [comment]
> Works very well. SO the last thing I expect to be fixed is _client-s IP in AUTH logs and/or user id in openvpn logs_ . It's not a show stopper but is important.

Just upgrade to lastest version. the client ip is part of the openvpn-auth-oauth2 logs

**Comment by aprudnev** [comment]
Confirming it. So we can link user's name and user's IP

For example

Mar  6 09:10:33 ***eisovpn03 openvpn-auth-oauth2[270174]:
time=2024-03-06T09:10:33.220-08:00 level=INFO msg="accept OpenVPN client
cid 2, kid 1" ip=*73.15.8.25:63003 <http://73.15.8.25:63003>* cid=2 kid=1
session_id=BLpbY6NdQa6ZSAB4 common_name=eqxeisovpn03 idtoken.subject=
*aroudnev* ...
Mar  6 09:10:38 ***eisovpn03 openvpn[270168]: eqxeisovpn03/*73.15.8.25:63003
<http://73.15.8.25:63003>* MULTI_sva: pool returned *IPv4=10.23.252.130*,
IPv6=(Not enabled)

TASK is done. IP/PORT define exact session on both sides. I can create a
filter which will post all connect/disconnects, with user IP and user name,
easily , with this.

**Comment by aprudnev** [comment]
I have a quick question - we must configure this parameter in the plugin

CONFIG_OAUTH2_ISSUER="https://test-sso.company.com"

(this is an example)

But if we read 'well known' json page (which is sp[ecified in , this page
PROVIDES this parameter. For example:

CONFIG_OAUTH2_ENDPOINT_DISCOVERY="
https://test-sso.company.com/.well-known/openid-configuration"

And this url contains issuer

"issuer": "https://test-sso.company.com",


(for example)

So it is not necessary, and if we check apache module or other
configuration, this parameter is not mandatory (as it usually comes
from a 'well-known' page).

Can it become optional (so it can be used if not provided by a
'well-known' page)?

**Comment by jkroepke** [comment]
The opposite the mandatory.

You have to define `CONFIG_OAUTH2_ISSUER="https://test-sso.company.com"` only. The underlying library will call `[issuer]/.well-known/openid-configuration`. That the standard OIDC discovery mechanism. Omit CONFIG_OAUTH2_ISSUER is not possible.

`CONFIG_OAUTH2_ENDPOINT_DISCOVERY` is only mandatory, if you have a crappy OIDC implementation,

**Comment by aprudnev** [comment]
Actually most OIDC implementations use *DISCOVERY url and do not request
ISSUER at all.

This is because (according to some forums, I did not check OIDC specs):

*The issuer value itself does not need to resolve to any resource or
anything online. Only if you want to conform to the Discovery spec
then <issuer>/.well-known/openid-configuration should resolve to the
Discovery metadata document that describes the Provider.*

So, while, if issuers is configured and discovery is not, you may expect
discovery as  *<issuer>/.well-known/openid-configuration *, it is not
guaranteed. What is guaranteed is that discovery page provides real issuer
value. I can see a few examples when discovery page is not formed this way

For example here issuer can not be used as base URL:

https://help.akana.com/content/current/cm/api_oauth/oauth_discovery/m_oauth_getOpenIdConnectWellknownConfiguration.htm

De facto ONE OF - DISCOVERY or ISSUER - is mandatory, and ISSUER can be
used alone ONLY if ISSUER is resolved into URL used for DISCOVERY page
(which, btw, is not always true).

PS. But I checked few our OIDC IDP and I agree, they all satisfy to this
requirement ( *<issuer>/.well-known/openid-configuration ). *For example
both PingFederate and PingOneforEnterprise satisfy it (last one use issuers
like this one -

CONFIG_OAUTH2_ENDPOINT_DISCOVERY="
https://sso.connect.pingidentity.com/xxxxe07e-7fe8-4dc8-b8b4-c70f023ddc7b/.well-known/openid-configuration
"
CONFIG_OAUTH2_ISSUER="
https://sso.connect.pingidentity.com/xxxxe07e-7fe8-4dc8-b8b4-c70f023ddc7b"
)

So you are correct in MOST (but not all) cases. And if a DISCOVERY page is
provided, the issuer must be taken from it.

As an example, see apache OIDC module (it was written by PING developers
and is extremely solid)

########################################################################################
#
# OpenID Connect Provider
#
# For configuration of a single static provider, not using OpenID Connect
Provider Discovery.
#
########################################################################################
# URL where OpenID Connect Provider metadata can be found (e.g.
https://accounts.google.com/.well-known/openid-configuration)
# The obtained metadata will be cached and refreshed every 24 hours.
# If set, individual entries below will not have to be configured but can
be used to add
# extra entries/endpoints to settings obtained from the metadata.
# If OIDCProviderMetadataURL is not set, the entries below it will have to
be configured for a single
# static OP configuration or OIDCMetadataDir will have to be set for
configuration of multiple OPs.
#OIDCProviderMetadataURL <url>

# OpenID Connect Provider issuer identifier (e.g. https://localhost:9031 or
https://accounts.google.com)
# Used when OIDCProviderMetadataURL is not defined or the metadata obtained
from that URL does not set it.
#OIDCProviderIssuer <issuer>

**Comment by jkroepke** [comment]
> Actually most OIDC implementations use *DISCOVERY url and do not request ISSUER at all.

Any major cloud like AWS; GCP; Azure requested an issuer. Kubernetes, the de-factor orchestrator for container workload is requested an issuer as well. Here are other examples as well:

* https://doc.traefik.io/traefik-enterprise/middlewares/oidc/
* https://dexidp.io/docs/connectors/oidc/

The [OIDC discovery spec](https://openid.net/specs/openid-connect-discovery-1_0.html) is very clear here: 

> The issuer value returned MUST be identical to the Issuer URL that was used as the prefix to /.well-known/openid-configuration to retrieve the configuration information.

It's important that a issuer is given first, because the client has to validate, if the same issue value is return back as given before.

In short: The discovery URL should be built from the Issuer, anything else is against the spec.

Nevertheless, any OIDC/OAUTH2 implementation is handled by https://github.com/zitadel/oidc/. The library is strict in terms of spec. As library consumer, I'm unable to change the behavior of library here.

**Comment by aprudnev** [comment]
It looks strange, but we configured dozens of clients, and all REQUIRED
metadata URL and not an issuer. It looks as specs added 'issuer to
metadata' rule later so older implementations did not use it.

In any case, it is not a problem, but I would state in documentation that
discovery is created from the issuer if not specified explicitly (if it is
not saying it already).

And thanks for this information, it was something new to learn.

**Comment by aprudnev** [comment]
Advanced template for APPROVED / DENIED page attached.

(company name and SLO URL must be adjusted)

We released system to BETA testing in production usage (it looks very
stable, and it's combination with Ping One Dock and all proxies on the same
host allows to create systems in a mouse click). We run few services per
OpenVPN server, generate profiles and publish them, proxy plugin, all on
the same URL (profiles are SSO protected).

The only drawbacks we found was _OpenVPN connect 3.4 do not like to
reconnect automatically, if it requests key's recreation_ (but it is
documented), and no reasonable client exists for Linux. The rest works
really great. I posted decorated template which allows to add disclaimer to
the success page and looks more professional (and have css inline, as
original template used external css which did not allow to make changes).
In addition, as clients do not allow to end session (they allow to
disconnect but SSO session will exists and you will connect automatically
on next click), we added SLO Logout link to this page, just in case.
